### PR TITLE
Avoid reporting disconnections for CancellationToken mismatch

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/Workspaces/Remote/ServiceHub/Services/BrokeredServiceBase.cs [hunk 1]
```diff
@@ -86,7 +87,7 @@ protected async ValueTask<T> RunServiceAsync<T>(Func<CancellationToken, ValueTas
         protected async ValueTask RunServiceAsync(Func<CancellationToken, ValueTask> implementation, CancellationToken cancellationToken)
         {
             WorkspaceManager.SolutionAssetCache.UpdateLastActivityTime();
-            using var _ = LinkToken(ref cancellationToken);
+            using var combined = cancellationToken.CombineWith(ClientDisconnectedSource.Token);
```
Comment (sharwell) [inline src/Workspaces/Remote/ServiceHub/Services/BrokeredServiceBase.cs:90]: ⚠️  The code below is expected to pass `combined.Token` to `implementation`, but instead passed `cancellationToken`. I'm going to fix this in a follow-up so I don't reset this build over it.

File: src/Workspaces/Remote/ServiceHub/Services/BrokeredServiceBase.cs [hunk 2]
```diff
@@ -71,7 +72,7 @@ protected Task<Solution> GetSolutionAsync(PinnedSolutionInfo solutionInfo, Cance
         protected async ValueTask<T> RunServiceAsync<T>(Func<CancellationToken, ValueTask<T>> implementation, CancellationToken cancellationToken)
         {
             WorkspaceManager.SolutionAssetCache.UpdateLastActivityTime();
-            using var _ = LinkToken(ref cancellationToken);
+            using var combined = cancellationToken.CombineWith(ClientDisconnectedSource.Token);
```
Comment (tmat) [inline src/Workspaces/Remote/ServiceHub/Services/BrokeredServiceBase.cs:75]: `combined` is not used? Isn't this a no-op?

Comment (sharwell) [reply src/Workspaces/Remote/ServiceHub/Services/BrokeredServiceBase.cs:75]: ➡️ See https://github.com/dotnet/roslyn/pull/48179#discussion_r497159268

File: src/Workspaces/Remote/ServiceHub/Services/BrokeredServiceBase.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/Remote/ServiceHub/Services/BrokeredServiceBase.cs:90]: ➡️ now fixed

**Review by tmat — APPROVED** [review]

File: src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs [hunk 3]
```diff
@@ -113,20 +113,26 @@ public static ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(PipeRe
...
             catch (EndOfStreamException)
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
                 throw exception ?? ExceptionUtilities.Unreachable;
             }
+            finally
+            {
+                // Make sure to complete the copy and pipes before returning, otherwise the caller could complete the
+                // reader and/or writer while they are still in use.
```
Comment (sharwell) [inline src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs:133]: @tmat This failure was caused by `ReadData` knowing _exactly_ how much data it needed, and stopping as soon as that was reached. At that point, this method returned even though the copy operation might not have processed the pipe completions. The caller completed them in the wrong order while the copy was still in progress, and produced an exception indicating something like this:

* The reader cannot be used after completed
* The writer cannot be used after completed

Propagation of the exception had unknown impact on the system, but may have been indirectly responsible for the TaskCanceledException.

Comment (tmat) [reply src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs:133]: I see.

**Pull request closed without merging**